WikiWatershed/model-my-watershed
# Adjust draw panel behavior

## Overview

This PR makes it so the draw panel behaves according to the new design in #1835. It also refactors the code in a few ways. Previously, the code for each draw tool (eg. Select by boundary, draw area, etc) spanned the `DrawWindow` view and a view for that particular tool. Now, the code for each draw tool is encapsulated in a separate view, and `DrawWindow` simply aggregates the set of draw tools. In addition, since 'Select by boundary', 'Draw area' and 'Delineate watershed' all have similar functionality, their corresponding views are derived from a base class, `DrawToolBaseView`. Some other dead and unnecessary code was also removed.

Connects #1835

### Demo

![screen shot 2017-04-27 at 6 01 44 pm 2](https://cloud.githubusercontent.com/assets/1896461/25506699/e1f44578-2b75-11e7-8e1a-2ac2ab286223.png)

### Notes

* @jfrankl will fine-tune the CSS as part of https://github.com/WikiWatershed/model-my-watershed/pull/1844
* As noted in #1827, when canceling a draw tool item, sometimes there is an erroneous error or warning message.

## Testing Instructions

* Click on a draw tool. Hover over the info icon to see more info and then select an item. 
* A turquoise box should appear beneath the dropdown with info about the item. Click the X to deselect that item.
* Click an item and then click another item, which should replace the previously selected item.
* Start using the item to select an area, but then switch to a different item. 
* Also try clicking the X for the item while in the middle of using it.
* Try using the item to actually select an area on the map and then run the analysis.
* Repeat for all the different draw tools. Note that the 'Upload file' tool does not have a dropdown, and clicking it directly opens the upload box.
